### DISCUSSION TITLE
TODOs | Future features

### DISCUSSION THREAD

**Discussion by GiorgosXou** [unicurses/unicurses#38] [Ideas]
*The place where we will discuss about future features a.k.a. TODOs*

**Comment by GiorgosXou** [comment]
Answers to @ttoommxx - [`TODO.md`](https://github.com/unicurses/unicurses/blob/2324548600d77a9a96952e0ddac24f39bcb359a1/TODO.md):

> Turning wch and wstr into ch and str and keep those. Strings in python have utf-8 encoding anyway, it does not makes much sense to keep every standard library. The official curses module for Python only supports addch and addstr

This will result in confusion. `w` stands for wide-characters, if you try to print non-ascii characters *(eg. emojis)* with `addch` it won't work. 

> change cursyncup to wcursyncup
change noutrefresh to wnoutrefresh

What do you mean?

> in general, change some names to be more faithful to python.curses and more consistent to prefix=w <-> affect input-specified windows

As I've said before "unicurses is trying to be as pure\functional-oriented\close-to the original C implementation, as a way of fast prototyping". But, we could try to make a pythonic-unicurses version, too. would you be interested?

> CCHAR: why do we let users pass in the number (ord) of the ascii character? CCHAR gets only either called by user, or called within the code with strings, not integers. Moreover ord of bytestring and ord of string return the same value, so I think we should just return ord or raise a custom error if ord fails, and avoid type checking

Not sure why, [this option](https://github.com/unicurses/unicurses/blob/29e0504febf1899de56d07f0cb3ebf9777a12d0c/unicurses/__init__.py#L373) was there since @chtenb. So, I never touched it really. He might be able to explain.

> why is it necessary to raise an exception when importing ctypes as failing will raise the same exact exception type?

Again this was since @chtenb. My guess is, that the exception message is more clear to the end-user, than the original.

> wgetkey: either split this into mvwgetkey or remove all unnecessary mv before and make everything into one function optional positional arguments 

Not sure why but I know there's this issue https://github.com/unicurses/unicurses/issues/5 and that for some reason I had a comment there that said [`# NEEDS_CHECK?`](https://github.com/unicurses/unicurses/blob/c4dea77a04e116539d928056614b9966ff8d924f/unicurses/__init__.py#L1049C34-L1049C49) *(might have to do with windows? idk)*

> PDC_LEAVEOK: change leaveok so that it actually behaves like ncurses, not sure why we need to use PDC_LEAVEOK, and do we actually need curscr at all? This can still be called by the user anyway, so that is the need? I suggest we make a function that returns curscr and let eveything else depend on scr_id instead. Is it not how curses functions are designed to start with anyway? Also, curscr is in ncurses too!

I think you kind of fixed that, right? `curscr` you mean [this line](https://github.com/unicurses/unicurses/blob/3d8afe9e3a45ddf31d143fd5dca5e4c77456fee8/unicurses/__init__.py#L1314)? idk 🤷‍♂️

> make it easier for users to find attributes - for example make a function that prints them or add them to the wiki

Sure.

> color_content does not seemt to work

Never used it but i found [this](https://stackoverflow.com/questions/59202852/ncurses-color-contents-function-not-returning-correct-colour-intensities)

> some functions such as erasewchar and killwchar require a pointer to be passed. How about we "translate" this class of functions into functions that accept vectors as pointers, and store whatever they have to store into the first coordinate? This approach would make porting C code to Python really easy.

I never used them, I got no idea what they do lol.

> make so that, if you only use one scr, you DON'T NEED to store stdscr anywhere

You don't need to store `stdscr` everywhere, it is global, wdym?

**Reply by ttoommxx** [reply]
Thank you for creating this and passing through my huge pull.

I realised the reason why CCHAR behaves that way is just because chars and int are essentially just numbers in C - for example ord('a')-'a'=0 - and so I suppose the author included that for portability reasons, as you pointed out multiple times.

As for how the module handles leave_ok checking on Windows, if it was for me I would remove PDC_LEAVEOK entirely and create a function that returns a pointer to curscr. I might be missing something and also don't have Windows for testing, so I would be happy to get feedback on this before doing so.

I created erasewchar and killwchar, the C functions need to be passed a pointer to a wchar and the characters are written therein. To emulate this behaviour, I decided to let the function accept a vector as input and write the characters in the first entry of such. Tested and working already.

**Reply by GiorgosXou** [reply]
> I realised the reason why CCHAR behaves that way is just because chars and int are essentially just numbers in C ...

True, thumbs up for the insight. Still though, I don't get why would someone use `CCHAR` with the `elif type(ch) == int`-condition in python. 

>  if it was for me I would remove PDC_LEAVEOK entirely... I might be missing something...

Thanks for bringing this to the surface, I think I figured out why, it's for cross compatibility. **If you check  [PDCurses](https://github.com/wmcbrine/PDCurses/blob/5c62af03e9a05e3a3ae8c8354c1234b772dcf4b0/pdcurses/getyx.c#L137) vs [Ncurses](https://github.com/mirror/ncurses/blob/87c2c84cbd2332d6d94b12a1dcaf12ad1a51a938/include/curses.h.in#L1154) implementation of `setsyx`**  you will see an `or` condition instead of `and`... which is a bit puzzling to me too, considering also what @wmcbrine said  [here](https://github.com/wmcbrine/PDCurses/pull/48#issuecomment-446449823). *(never tested it, but i guess that was the whole point of unicurses global variable PDC_LEAVEOK)*

> I created erasewchar and killwchar, the C functions need to be passed a pointer to a wchar and the characters are written therein. 

I guess something like what the [wgetstr](https://github.com/unicurses/unicurses/blob/3d8afe9e3a45ddf31d143fd5dca5e4c77456fee8/unicurses/__init__.py#L1301-L1303) does but for `wchar_t`, right?

**Reply by ttoommxx** [reply]
Alright. Will leave leaveok there, I am sure the author knew what they were doing + I can't test in on Windows either. I did change CCHAR to isinstance and check for bytes and str instead just str, as bytes are actually char* in C. Will keep testing the functions before the new commit!

**Comment by chtenb** [comment]
All the stuff from commit https://github.com/unicurses/unicurses/commit/eb2b100fb2e29ebece9e8961bca788d1e6447e0b was inherited from Michael Kamensky

**Reply by GiorgosXou** [reply]
lol, I thought you were Michael Kamensky. who's Michael Kamensky?

**Reply by chtenb** [reply]
😄  No idea, hadn't been able to contact him at the time. The project was abandoned at sourceforge.

**Reply by GiorgosXou** [reply]
Mikhail Vasilievich Kamensky, this guy authored a [paper](https://www.researchgate.net/publication/351650949_The_Role_Of_Ethnopsychology_In_Regulating_Culture-Based_Language_Acquisition_Techniques) in 2020... interesting. But anyways.. lol 🤷‍♂️

**Reply by chtenb** [reply]
https://sourceforge.net/projects/pyunicurses/

**Comment by GiorgosXou** [comment]
**TODO:** `cchar_t` on windows  + add [`wborder_set`](https://stackoverflow.com/a/69580875/11465149) + replace substitute functions like `mvwadd_wch` and stuff..
https://github.com/wmcbrine/PDCurses/blob/5c62af03e9a05e3a3ae8c8354c1234b772dcf4b0/curses.h#L90-L98

**Reply by ttoommxx** [reply]
Can you clarify what you mean by "replace substitute functions like mvwadd_wch and stuff.."?

**Reply by GiorgosXou** [reply]
For example, on windows [we use a bunch of functions for `mvwadd_wch`](https://github.com/unicurses/unicurses/blob/f0ecaf9e5d016c3d4ba386391d6b751073a93f29/unicurses/__init__.py#L1670-L1679) instead of just using a single `lib1.mvwadd_wch`. This is a result of me not knowing exactly how to handle `cchar_t` for windows.

**Reply by GiorgosXou** [reply]
Additionally *(not tested but)* I suspected that `wborder_set` with `lib1.wattrset` might not work as expected.... but at the end of the day it might not be necessary at all and my assumptions might were fault.

**Reply by ttoommxx** [reply]
Oh yes, that makes sense. Was also trying to port a ncurses/PDcurses compatible function but had to stop when I saw was using this cchar_t type, will also try to investigate workarounds :) As for `wborder_set` I did not see it in the repo, are you talking about implementing it or something else?

**Reply by GiorgosXou** [reply]
> are you talking about implementing...

Yep (:

**Comment by GiorgosXou** [comment]
Regarding [this PR](https://github.com/unicurses/unicurses/pull/42) It made me think if: 

1.  `getstr`, `wgetstr`, `mvgetstr`, `mvwgetstr` etc. should have an additional `buffer_size=1023` variable
2.  All the  `getnstr`, `wgetnstr`,  ... `n`-etc. functions should have a `ctypes.create_string_buffer(n)` instead of `1023` *(although it should be tested if both ncurses,pdcurses handle 2-byte chatacters [such as: emojis or έϊά] as `2`-n or `1`-n characters and so on)*
